src/home_robot/home_robot/perception/detection/grounded_sam/sam_perception.py

```diff
@@ -0,0 +1,172 @@
+             list of classes (as a single string)
+            sem_gpu_id: GPU ID to load the model on, -1 for CPU
+            checkpoint_file: path to model checkpoint
+            verbose: whether to print out debug information
+        """
+
+        # Building MobileSAM predictor
+        checkpoint = torch.load(checkpoint_file)
+        self.mobile_sam = setup_model()
+        self.mobile_sam.load_state_dict(checkpoint, strict=True)
+        device = gpu_device_id
+        if device is None:
+            device = DEVICE
+        self.mobile_sam.to(device=device)
+        self.custom_vocabulary = custom_vocabulary
+        self.sam_predictor = SamPredictor(self.mobile_sam)
+        final_mask_generator_kwargs = copy.deepcopy(_DEFAULT_MASK_GENERATOR_KWARGS)
+        final_mask_generator_kwargs = dict_update(
+            final_mask_generator_kwargs, mask_generator_kwargs
+        )
+        self.mask_generator = SamAutomaticMaskGenerator(
+            model=self.mobile_sam, **final_mask_generator_kwargs
+        )
+
+    def reset_vocab(self, new_vocab: List[str]):
+        """Resets the vocabulary of Detic model allowing you to change detection on
+        the fly. Note that previous vocabulary is not preserved.
+        Args:
+            new_vocab: list of strings representing the new vocabulary
+            vocab_type: one of "custom" or "coco"; only "custom" supported right now
+        """
+        self.custom_vocabulary = new_vocab
+
+    def generate(self, image):
+        return self.mask_generator.generate(image)
+
+    def predict(
+        self,
+        obs: Observations,
+        depth_threshold: Optional = None,
+        draw_instance_predictions: bool = False,
+    ):
+        """
+        Arguments:
+            obs.rgb: image of shape (H, W, 3) (in RGB order)
+            obs.depth: depth frame of shape (H, W), used for depth filtering
+            depth_threshold: if specified, the depth threshold per instance
+
+        Returns:
+            obs.semantic: segmentation predictions of shape (H, W) with
+             indices in [0, num_sem_categories - 1]
+            obs.task_observations["semantic_frame"]: segmentation visualization
+             image of shape (H, W, 3)
+        """
+
+        if draw_instance_predictions:
+            raise NotImplementedError
+        # Predict classes and hyper-param for GroundingDINO
+        # CLASSES = self.custom_vocabulary
+        height, width, _ = obs.rgb.shape
+
+        # convert to uint8 instead of silently failing by returning no instances
+        image = obs.rgb
+        if not image.dtype == np.uint8:
+            if image.max() <= 1.0:
+                image = image * 255.0
+            image = image.astype(np.uint8)
+
+        self.mask_generator.generate(image)
+
+        # semantic_map, instance_map = overlay_masks(
+        #     detections.mask, detections.class_id, (height, width)
+        # )
+
+        # obs.semantic = semantic_map.astype(int)
```

Comment:
dead code?

Reply:
Not sure what we want this interface to actually be?
The segmenter returns masks -- not an instance image. I'm leaving this code to indicate what the "usual" interface is, but we can remove it